### ISSUE TITLE
-Wenum-constexpr-conversion should be a hard error, not a downgradable error

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#59036]
The `-Wenum-constexpr-conversion` warning was created to account for the fact that casting integers to enums outside of the valid range of the enum is UB in C++17. Constant expressions invoking UB lead to an ill-formed program.

The current behavior of Clang is that it does warn and it does error, but the message is nevertheless a warning that the user can easily suppress via `-Wno-enum-constexpr-conversion`.

Since the program is **ill-formed**, I don't think this should be a warning that can be disabled. Just like one can't disable the warning in this ill-formed code:

```cpp
constexpr int x[10]{};
constexpr int y = x[10];
```

Therefore I propose to make the diagnostic a hard error that cannot be disabled. Additionally, this "hard error" behavior should only be active in C++17 (it's currently active in for any standard).

The warning could be repurposed to be a general warning that one can enable, and is active not only in `constexpr` expressions, but in any expression that involves casting an integer to an enum.

**Comment by thesamesam** [comment]
https://clang.llvm.org/docs/ReleaseNotes.html#potentially-breaking-changes (for Clang 16) says it's an error by default but is currently downgradable (until Clang 17). I guess this bug is to track remembering to do so in Clang 17?

**Comment by carlosgalvezp** [comment]
Nice, I wasn't aware of that, thanks! Yes that makes a lot of sense, thanks for pointing out. Let's keep this bug as a reminder. I will open a new ticket for having a warning that works also on non-constexpr contexts.

**Comment by tbaederr** [comment]
Did this happen?

**Comment by carlosgalvezp** [comment]
https://reviews.llvm.org/D150226

**Comment by carlosgalvezp** [comment]
@AaronBallman @dwblaikie We are now on Clang 19, do you think it's the right time to pull the trigger and turn this into a non-downgradable error? It has been a warning also in system headers since Clang 18.

**Comment by dwblaikie** [comment]
Looks like binutils is still suppressing this warning: https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=include/diagnostics.h;h=8cc2b493d2c02ba7dbc5879207746107ad7143a0;hb=refs/heads/master#l81

As is openmp in the LLVM project itself (https://github.com/llvm/llvm-project/blob/58c494f47cf56a30bf50024d22661c75003db809/openmp/cmake/HandleOpenMPOptions.cmake#L44)

And I also found this `magic_enum` project disabling it too: https://github.com/Neargye/magic_enum/blob/b4ada55c8cea21f3c73c5a58cebd368a227465a1/include/magic_enum/magic_enum.hpp#L71

At least for the openmp usage and probably the binutils usage, it'd be nice to see if some progress can be made removing those warning suppressions.

**Comment by carlosgalvezp** [comment]
FYI I'm bringing this issue to the binutils project.
https://sourceware.org/bugzilla/show_bug.cgi?id=31331

I agree with Aaron, it's way easier to just suppress the warning and call it a compiler bug than actually do some research and understand the subtle issues when doing these tricks with enums. I think that the more we delay it, the more projects will suppress the warning as they bump their clang installation, and so the harder it will be for us to introduce it.

That said, I'm not opposed to waiting another release or two.

**Comment by RJVB** [comment]
> The `-Wenum-constexpr-conversion` warning was created to account for the fact that casting integers to enums outside of the valid range of the enum is UB _in C++17_. 

Quick question: it is thus *not* UB in earlier C++ dialects (or C)? I've just come across this "werror" being triggered building Phonon code as `gnu++11`with clang-16 .

**Comment by carlosgalvezp** [comment]
Technically speaking it's _unspecified_  behavior until C++17, after that it's UB.

**Comment by RJVB** [comment]
>Technically speaking it's _unspecified_  behavior until C++17, after that it's UB.

And that's better or worse how? :)

Probably not the place, but as a mostly-C programmer I fail to see why it would be either from an implementation point of view. An enum type is some kind of integer behind the scenes (a standard `int` IIRC) and as long as the "incriminated" values you try to assign a variable of the enum type in question are within the range of that type there shouldn't be any ambiguity about the result of an assignment? (And assigning a value that can't be represented by the underlying type could give just the same warning as other assignments of that kind?!)

Case in point: the "faulty" Phonon code that led me here could be "fixed" by just adding `: unsigned` to the enum declarations. Seeing that it made sense that this "error" had never caused any runtime problems...

**Comment by RJVB** [comment]
>Technically speaking it's _unspecified_  behavior until C++17, after that it's UB.

Oh, and doesn't this imply that clang shouldn't raise an error about it until C++17,  technically speaking of course?

**Comment by carlosgalvezp** [comment]
> And that's better or worse how? :)
Worse, because with UB "anything can happen", there's no restrictions. With Unspecified Behavior, only a defined set of behaviors is allowed (but which one is used is not specified in the Standard).

> why it would be either from an implementation point of view. 

People writing the Standard should probably answer that :) Some things are kept for historical reasons/backwards compatibility, even though they might make little sense today.

> a standard `int` IIRC

That's true in C, but not in C++, where there's quite a few cases depending on whether the enum is a C-style enum, scoped/unscoped, with explicit underlying type, etc.

> Seeing that it made sense that this "error" had never caused any runtime problems...

It's often the case with UB that the program "just works", but it can break anytime in the future when refactoring code, enabling optimizations, updating the compiler, etc. 

> clang shouldn't raise an error about it until C++17,  technically speaking of course?

Good question, I'm not sure actually. The Standard requires that it's an error from C++17, but nothing stops an implementation to make it an error earlier than that. Maybe @AaronBallman @dwblaikie know better?

**Comment by RJVB** [comment]
On Saturday April 27 2024 12:03:21 Carlos Galvez wrote:

>> a standard `int` IIRC
>
>That's true in C, but not in C++, where there's quite a few cases depending on whether the enum is a C-style enum, scoped/unscoped, with explicit underlying type, etc.

And that hasn't evolved more or less slowly from "true in C" (at some point C++ was just a preprocessor)?

>> Seeing that it made sense that this "error" had never caused any runtime problems...
>
>It's often the case with UB that the program "just works", but it can break anytime in the future when refactoring code, enabling optimizations, updating the compiler, etc. 

True, and it's quite possible the latent bugs I ever found in my own code by simply building it on a different platform (OS) were examples. (But to remain with my own example at hand: 2 compilers on 2 different platforms generate exactly the same code, shared libraries, with and with the "unsigned" fix. I guess "unspecified" doesn't mean that there can't be a sensical thing to do...)

>Good question, I'm not sure actually. The Standard requires that it's an error from C++17, but nothing stops an implementation to make it an error earlier than that.

But one that can be overridden, in that case? Standards are probably akin to legislation, where most things are either explicitly forbidden or implicitly allowed?

**Comment by carlosgalvezp** [comment]
> And that hasn't evolved more or less slowly from "true in C" (at some point C++ was just a preprocessor)?

C++ is a different language than C, with different standards. Sure, there is a lot of backwards compatibility with C, but there are slight differences.

> I guess "unspecified" doesn't mean that there can't be a sensical thing to do...

Absolutely, but there may be multiple sensical things to do and different implementations don't necessarily need to agree.

> But one that can be overridden, in that case? Standards are probably akin to legislation, where most things are either explicitly forbidden or implicitly allowed?

Currently the warning can be suppressed in order to give time to people to update their code, before making it a hard error. In that sense Clang is currently not complying with the Standard (but the goal is to make it be).

Standards describe mostly what is allowed, and often clarify that "otherwise, the behavior is undefined".

**Comment by RJVB** [comment]
On Sunday April 28 2024 07:05:25 Carlos Galvez wrote:
>C++ is a different language than C, with different standards. Sure, there is a lot of backwards compatibility with C, but there are slight differences.

Of course it is.

>Currently the warning can be suppressed in order to give time to people to update their code, before making it a hard error. In that sense Clang is currently not complying with the Standard (but the goal is to make it be).

No argument against making it a hard error for C++17 and newer; I *was* talking about earlier standards. I'll admit that I can't really wrap my head around whether or not "updating the code" can always be as simple as in the single example I've seen, or less kludgy than just replacing the offending enum by a regular int data type.

**Comment by carlosgalvezp** [comment]
I suppose that should be technically possible. What's the argument for keeping it since C++11 instead of C++17 @AaronBallman @dwblaikie ?

**Comment by AaronBallman** [comment]
> I suppose that should be technically possible. What's the argument for keeping it since C++11 instead of C++17 @AaronBallman @dwblaikie @shafik ?

This was a change as a result of a [defect report](https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#2338) and we typically apply defect reports as though the original specification was always worded with the fix applied.

**Comment by carlosgalvezp** [comment]
Thanks for the explanation! I guess you meant this one? 
https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#1766

In cppreference it even says it applies from C++98 onwards.

@RJVB does this answer your question?

**Comment by RJVB** [comment]
>@RJVB does this answer your question?

Not really, as far as I can understand such "philosophical" reasonings I still seem to see retro-active changing of old standards and deliberate breaking of backwards compatibility. Two things I have difficulties wrapping my head around because so contrary to the everything I ever did.

**Comment by AaronBallman** [comment]
The purpose of the committee adopting a change as a defect report is to fix bugs with features of the language or library. While that can and does change behavior previously promised by the standard, the intent is that the previous standard *was wrong* and the changes are fixing an issue.

Whether that's a good idea or not is a matter for debate, but it is a practice that WG21 has had for over a decade and is something the committee strongly encourages implementations to do so that the language remains coherent. We are not required to implement a defect report in all language modes in terms of meeting conformance requirements for a standard, but it is a tradeoff. Implementing DRs the way the committee requests often helps migration between standards modes, but at the expense of hurting migration between compiler versions, unfortunately.

In this case, we've signaled our intent to turn this into a hard error in a future release of Clang. If there's compelling rationale for why we should not turn it into a hard error, it's something we'll certainly consider when deciding when or if to follow through on that intent.

> Case in point: the "faulty" Phonon code that led me here could be "fixed" by just adding `: unsigned` to the enum declarations. Seeing that it made sense that this "error" had never caused any runtime problems...

C and C++ differ in their treatment of enumerations. In C before C23, the enumerators were of type `int` and the underlying type of the enumeration was either `char`, a signed integer type, or an unsigned integer type (implementation-defined). A lot of implementations had extensions to these rules, and those extensions were largely ratified for C23. However, in all versions of C++, an enumeration's underlying type was the smallest power-of-two integer bit-field that could represent all the values of the enumeration (up to a max of `unsigned int`). So: `enum E { zero, one, two, three };` in C would be able to hold values that can fit in an `int` while in C++ it would be able to hold values up to `unsigned int : 2`. It's not possible to say "this makes sense in C so it should be fine in C++ as well" in all cases, and this is one such case. Adding an explicit underlying type to the enum declaration in C++ forces the underlying type to be large enough to hold any value that fits within that type, that's why it fixes the incorrect code. The code may not have caused runtime problems for you, but it was still undefined behavior and not something you could rely on working correctly across compilers or compiler upgrades. It sounds like this diagnostic caught a real issue that would have otherwise remained latent in the code base and could have caused a hard-to-debug issue if an optimizer started taking more advantage of the C++ semantics for enumerations.

**Comment by RJVB** [comment]
On Tuesday May 07 2024 07:03:53 Aaron Ballman wrote:
>The code may not have caused runtime problems for you, but it was still undefined behavior and not something you could rely on working correctly across compilers or compiler upgrades.

According to what was said above it was UB only from C++17 onwards, while the code in question probably predates even C++11. Generations of compilers did not miscompile it (or it would have been changed before) and IMHO it would be a conscious choice to start miscompiling it. I would find it very hard to accept a breaking change in what is arguably a compatitibity mode (`-std=c++11`) nowadays, and if I hadn't decided to take the time to understand the very cryptic error messages I might have concluded that I'd best continue to use older (and generally faster) clang versions, or GCC (also generally faster nowadays...) I recall trying to understand the messages and thinking that some really looked like one of those situations where the compiler somehow can't see that you're NOT using a non-initialised variable.

>It sounds like this diagnostic caught a real issue that would have otherwise remained latent in the code base and could have caused a hard-to-debug issue if an optimizer started taking more advantage of the C++ semantics for enumerations.

Well, as I said, the message (also when printed as a warning) was certainly no help in debugging the issue, which could well explain why people have the reflex to just make the warning really only a mere warning. Build the same code with g++'s `-Wconversion` and you get something much more useful that actually pinpoints the place where you can fix the problem. That's how I found the location in Phonon that hadn't (possibly still hasn't) been fixed yet.

**Comment by carlosgalvezp** [comment]
Fyi I have now put up a patch for review to the binutils project: 
https://sourceware.org/pipermail/gdb-patches/2024-May/209513.html

**Comment by AaronBallman** [comment]
> We are now at Clang 20. Would it be now a good time to turn the warning into a hard error?

I think it's worth a shot, yes.

> I'm pinging the GDB community but I get no reviews of the above patch. Unfortunately there's not much more I can do. Hopefully if we turn the warning into an error it will put some pressure into fixing the code with the available patch.

Thank you for keeping on top of that!

**Comment by carlosgalvezp** [comment]
Thanks! I can prepare a patch. For consistency, I think it would be nice to show the error like this:

```
<source>:2:15: error: constexpr variable 'a' must be initialized by a constant expression
    2 | constexpr int a  = x[15];
      |               ^    ~~~~~
<source>:2:20: note: cannot refer to element 15 of array of 10 elements in a constant expression
    2 | constexpr int a  = x[15];
      |                    ^
```

I.e. first print the generic error, and follow-up with the `note` (which is the current diagnostic as a warning). This clarifies that the hard error is only in the context of a constant expression.

**Comment by mborgerding** [comment]
What is the benefit to removing the "downgrade" that could outweigh the pain this causes by breaking years-old code?

AFAIK there is currently no released version of boost that [can be compiled without `-Wno-enum-constexpr-conversion`](https://godbolt.org/z/PGYG5ezGW).  Even after such a fix is released (slated for 1.87), it will be years before that boost reaches stable package managers.

**Comment by RJVB** [comment]
> Even after such a fix is released (slated for 1.87), it will be years before that boost reaches stable package managers.

Just build it with G++ I suppose? Mostly a good trade-off anyway from what I've seen ;)

**Comment by carlosgalvezp** [comment]
> AFAIK there is currently no released version of boost that

The example you provided [compiles without errors](https://github.com/boostorg/mpl/pull/77) on Boost 1.81 (released December 2022) and clang trunk.

The issue was [fixed](https://github.com/boostorg/mpl/pull/77) on June 24th, and Boost release 1.86 has been available since August 14th.

**Comment by carlosgalvezp** [comment]
The benefit is that it shows that the program is ill-formed (as per the standard) - it's not a _warning_, it's a hard compile error, just like forgetting a semicolon. 

If an escape hatch is wanted, I believe it should be implemented as an `-f` flag instead, like it's done for e.g. `-fno-strict-aliasing`, `-fwrapv`, etc.

**Comment by mborgerding** [comment]
I like the idea of the `-f` flag.

Since the error has been historically unenforced, many code bases are now broken, apparently by the new compiler (not really, but that's how it is may be perceived).

**Comment by carlosgalvezp** [comment]
> I like the idea of the -f flag.

Any thoughts @AaronBallman ?

**Comment by carlosgalvezp** [comment]
> Since the error has been historically unenforced, many code bases are now broken, 

Also, this is exactly what happens with UB, one can't really blame the compiler for that. The same happened when GCC "broke" lots of code when optimizing out integer overflow checks.

**Comment by mborgerding** [comment]
Is there an optimization to be had in this case?

**Comment by carlosgalvezp** [comment]
No, since this error applies only to constant expressions.

**Comment by AaronBallman** [comment]
> > I like the idea of the -f flag.
> 
> Any thoughts [@AaronBallman](https://github.com/AaronBallman) ?

`-f` flags introduce a new language dialect, so there would need to be a pretty strong reason for adding it. I don't see the need here yet.

> But ... it still doesn't answer my question. What is the benefit to removing the -Wno-enum-constexpr-conversion downgrade?

One of the key benefits of constant expressions is that their evaluation *cannot* contain undefined behavior. For example, if designing a fully constexpr interface, you can test that interface using `static_assert` and if the code compiles, it is correct and is known to have no undefined behavior (at least with the test inputs). That particular enum conversion is undefined behavior so allowing it in constant expressions means users cannot actually rely on that core property.

The ability to downgrade this error into a warning was only ever intended as a stopgap solution for people tracking the latest version of Clang (to ease their upgrade path). We gave folks about two years to react and then removed the stopgap.

**Comment by mborgerding** [comment]
Since the UB is only with c++17, does that mean compiling for previous standards will still allow the workaround?

**Comment by carlosgalvezp** [comment]
No, this is applied from C++11 (I think), since it's a Defect Report. See previous response: 
https://github.com/llvm/llvm-project/issues/59036#issuecomment-2098487902

**Comment by mborgerding** [comment]
I'm not a contributor to this fine project, so my influence is understandably small.  If the LLVM maintainers value purity over usability and practicality, I respect that.  Take the high road.  It's your project.  It's your passion. You decide your goals.

But I don't agree with breaking years-old code for pedantic adherence to the c++17 standard (even applied back to c++11 compilation ?), __without any workaround__.

It sounds like I need to caution my customers who build on RedHat Linux or older Ubuntu LTS that the workaround will stop working soon and they should avoid using upcoming LLVM compilers unless they are also willing to upgrade other required libs beyond the package manager.

**Comment by RJVB** [comment]
On Tuesday December 03 2024 07:27:27 Carlos Galvez wrote:
>No, this is applied from C++11 (I think), since it's a Defect Report. 

That seems like an arbitrary cut-off then?

>See previous response: https://github.com/llvm/llvm-project/issues/59036#issuecomment-2098487902

Yes, this was discussed before but my opinion about it hasn't changed. Maybe if error reporting has improved but when I was confronted with a build failure (due to -Werror in my case) it was GCC that allowed me find and fix the culprit).

**Comment by AaronBallman** [comment]
> I'm not a contributor to this fine project, so my influence is understandably small. If the LLVM maintainers value purity over usability and practicality, I respect that. Take the high road. It's your project. It's your passion. You decide your goals.

We're following the guidance from the standards committee; they thought this issue was sufficiently important to change its behavior as a defect report: 

https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#1766
https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#2338

> But I don't agree with breaking years-old code for pedantic adherence to the c++17 standard (even applied back to c++11 compilation ?), without any workaround.

We're not breaking years-old code; we're diagnosing code that's been broken for years and was hoping the behavior from the compiler was reasonable. Enumerations in C++ are not the same as enumerations in C in that C++ enumerations use a notional power-of-two integer type as the underlying type of an unscoped enumeration, not `int` as in C. And compilers can (and do) take advantage of this outside of constexpr contexts: https://gcc.godbolt.org/z/evPx7oYe4

(There is an interesting question there of whether this diagnostic should be tied to `-fstrict-enums` or not. Currently it's not.)

> It sounds like I need to caution my customers who build on RedHat Linux or older Ubuntu LTS that the workaround will stop working soon and they should avoid using upcoming LLVM compilers unless they are also willing to upgrade other required libs beyond the package manager.

If the changes are sufficiently disruptive, we can certainly walk them back and leave it as an error which can be downgraded to a warning for another release or two. However, there's not been much indication yet that this is significantly disruptive in practice. Are you finding otherwise?

**Comment by RJVB** [comment]
On Tuesday December 03 2024 08:33:44 mborgerding wrote:
>It sounds like I need to caution my customers who build on RedHat Linux or older Ubuntu LTS that the workaround will stop working soon and they should avoid using upcoming LLVM compilers unless they are also willing to upgrade other required libs beyond the package manager.

They will presumable be using older hardware too? If so they'll probably see faster and less resource-hungry builds too by *not* using clang...

**Comment by mborgerding** [comment]
> However, there's not been much indication yet that this is significantly disruptive in practice. Are you finding otherwise?

Has the upcoming _unreleased_ version been disruptive in practice?  Shockingly no. ;)

Recent versions of intel LLVM-based icpx were slightly disruptive, but not too bad.  There was a compiler flag workaround.

**Comment by AaronBallman** [comment]
> Has the upcoming _unreleased_ version been disruptive in practice?

Correct. The unreleased version of the compiler gets nontrivial amounts of testing in practice, so if this broke something significant, we're likely to hear about it before the final rc's get made.

**Comment by carlosgalvezp** [comment]
> There is an interesting question there of whether this diagnostic should be tied to -fstrict-enums or not. Currently it's not.

That could be an option. It might however open the door to allowing similar things with similar flags, e.g. don't error on signed integer overflow in constant expressions with `-fwrapv`.

**Comment by AaronBallman** [comment]
> > There is an interesting question there of whether this diagnostic should be tied to -fstrict-enums or not. Currently it's not.
> 
> That could be an option. It might however open the door to allowing similar things with similar flags, e.g. don't error on signed integer overflow in constant expressions with `-fwrapv`.

Yeah, the behavior could end up inconsistent in ways that are similarly frustrating -- e.g., a library developer who compiles with the flag disabled (which it is by default) may not get any diagnostics, ship their library, and then a user who does enable the flag consumes the library header and gets errors they can't fix.

So in terms of options I see currently, we have:

1) status quo -- it's a hard error in Clang 20 in all C++ language modes, based on the rationale that we're following the committee's recommendations and we're diagnosing code that has undefined behavior in a constant expression, which is expected behavior for any UB in a constant expression.
2) don't implement as a DR -- it's still a hard error in Clang 20, but only in C++17 and up. In C++11/14, it would be a warning and users would lose the property of "no UB in a constant expression" in a known case. This would be based on user disruption from the changes. It may be seen as inconsistent by users because it's UB in all C++ language modes.
3) don't implement the changes at all -- it's back to being a warning only in Clang 20 in all language modes. This would be based on user disruption for the changes. It would be self-consistent, unfortunately diverge from what the standard requires (in a conforming way though), but it would still be more than any other C++ compiler does currently in terms of diagnosing the situation: https://godbolt.org/z/aP976TPe9 If we went with this option, I think we'd have to go back to WG21 with a paper explaining that the DRs resolved wrong and how we'd rather see them resolved, but I don't expect much appetite from WG21 for a change in this area (the only change I think the committee could make would be to stop using the power-of-two notional underlying type for unscoped enumerations, but that's been the definition in C++ since 1998 and there are compilers which take advantage of it via things like `-fstrict-enums`).

**Comment by RJVB** [comment]
On Wednesday December 04 2024 00:13:21 Carlos Galvez wrote:
>signed integer overflow in constant expressions with `-fwrapv`.

Doesn't overflow always simply discard the overflowing bits? Or are we still talking about C++ only here too, where apparently "thou shalt not depend on hardware tricks to implement a cheap self-resetting counter"?

**Comment by carlosgalvezp** [comment]
> So in terms of options I see currently

Personally I'm happy with 1. I think option 2 is probably a good tradeoff, but ultimately is just delaying the problem. 

> Doesn't overflow always simply discard the overflowing bits?

No, this is simply undefined behavior, both in C and C++. Check the following example: the compiler will always emit "false" regardless of the input:
https://godbolt.org/z/3sjYTb7Y9

**Comment by RJVB** [comment]
On Wednesday December 04 2024 04:43:11 Aaron Ballman wrote:
>2) don't implement as a DR -- it's still a hard error in Clang 20, but only in C++17 and up. In C++11/14, it would be a warning and users would lose the property of "no UB in a constant expression" in a known case. This would be based on user disruption from the changes. It may be seen as inconsistent by users because it's UB in all C++ language modes.

I know this was discussed before, but wasn't it declared UB only in C++17? That idea must come from somewhere, and clang++ has clearly managed to avoid miscompiling code offending this principle until now. Don't mistake "users" for compiler experts, I think the vast majority won't see an inconsistency when a standard is being followed to the letter (and not more than that).
After all the whole debate here is about old code that may not even follow C++11 .

You missed an option, I think:

4) drop support for obsolete C++ standards (arbitrarily defined as pre-C++17).
That would be perfectly consistent with LLVM's, erm, forward looking approach.

**Comment by AaronBallman** [comment]
> I know this was discussed before, but wasn't it declared UB only in C++17?

Yes and no. In C++14 and before, it was unspecified behavior rather than undefined behavior. Unspecified behavior means "you'll get a result but it doesn't have to make sense or be consistent", whereas undefined behavior means "you'll get what you get, no promises about anything". Both are indicative of the user's code being invalid.

Additionally, there's two ways issues get resolved in WG21: either as a Defect Report (DR) or not. If an issue is not resolved as a DR, then the behavior is expected to change in that release going forward. If an issue is resolved as a DR, then the behavior is expected to change in all language modes where the issue could appear. So this was declared UB in C++17, but as a DR, which means the committee would like us to treat it as though it was always undefined rather than unspecified.

> That idea must come from somewhere, and clang++ has clearly managed to avoid miscompiling code offending this principle until now. Don't mistake "users" for compiler experts, I think the vast majority won't see an inconsistency when a standard is being followed to the letter (and not more than that). After all the whole debate here is about old code that may not even follow C++11 .

There are natural tensions here. Some people want existing code to work at all costs. Some people don't care at all about existing code if they can prove that code is "broken" somehow. Most folks are somewhere in between those two extremes. WG21 wants a perfect language which never has bugs, users want their existing code to continue to work but are fine with behavioral changes so long as it makes things faster, and implementers have to find ways to work with both sets of desires.

> You missed an option, I think: 4) drop support for obsolete C++ standards (arbitrarily defined as pre-C++17). That would be perfectly consistent with LLVM's, erm, forward looking approach.

That is an option, but it's a non-starter because there is *far* too much old code that still needs to be compiled. For example, think about a Linux distribution and how many packages it builds -- many of those are C++98/03 and quite don't see regular updates, but still need to compile with the same compiler as other packages.

**Comment by RJVB** [comment]
On Wednesday December 04 2024 05:28:01 Carlos Galvez wrote:
>No, this is simply undefined behavior, both in C and C++. Check the following example: the compiler will always emit "false" regardless of the input:
>https://godbolt.org/z/3sjYTb7Y9

Seems to me that it shouldn't do anything else from a mathematical point of view ... I can't think of a finite value of x where `x+1` is not larger than `x` itself. But, C(++) isn't math...

However, that's only true for compiling with optimisation. Clang <= 12 (don't have any newer on the system I'm on) and GCC <=13 will have that `will_overflow` function return `true` for x=2147483647 . They will also do that *with* optimisation when using `const volatile y=x+1; return y < x;`.

FWIW, shouldn't at least GCC warn about the expression being UB when compiling with -Wsequence-point?

**Comment by mborgerding** [comment]
From my limited perspective ... my and my customers' needs would be addressed by the option that differentiates at c++17: 
* for pre 17: a dire warning about UB (not suppressible?)
* for post 17: a hard error

**Comment by RJVB** [comment]
On Wednesday December 04 2024 05:57:16 Aaron Ballman wrote:

>That is an option, but it's a non-starter because there is *far* too much old code that still needs to be compiled. For example, think about a Linux distribution and how many packages it builds -- many of those are C++98/03 and quite don't see regular updates, but still need to compile with the same compiler as other packages.

It was tongue in cheek of course, but...

Hah, consider how many people are also using older hardware that doesn't run a current version of whatever OS family they use (and how happily LLVM dropped official support for those) ...

How many Linux distributions even use clang as their system compiler? Plus, that code gotta be "fixed" sometime, no? O:^)
(I'm not really up to speed how C++ standards "work", but I'd presume they're incremental so most code that's a proper standard X will also conform to standard X+1?)

**Comment by AaronBallman** [comment]
> From my limited perspective ... my and my customers' needs would be addressed by the option that differentiates at c++17:
> 
>     * for pre 17: a dire warning about UB (not suppressible?)
>     * for post 17: a hard error

It would help me if I understood your needs better. I know what you want, but I don't know the severity of why you want it. Is there something preventing your customers from fixing their code? Is this impacting 1000s of customers? Something else?

**Comment by AaronBallman** [comment]
> It was tongue in cheek of course, but... Hah, consider how many people are also using older hardware that doesn't run a current version of whatever OS family they use (and how happily LLVM dropped official support for those) ... How many Linux distributions even use clang as their system compiler?

The Linux kernel itself, FreeBSD, Darwin, others...

> Plus, that code gotta be "fixed" sometime, no? O:^) (I'm not really up to speed how C++ standards "work", but I'd presume they're incremental so most code that's a proper standard X will also conform to standard X+1?)

The way I'd describe it is like this: The C committee works really hard to avoid breaking existing code when updating standards, but it still happens on occasion. The C++ committee knowingly breaks existing code when updating standards, but does so in an opinionated way that aims to limit the blast radius somewhat (e.g., they'll break "bad", "ugly", or "overly clever" code, but they won't break "important" code, where all of these are value judgements).

**Comment by AaronBallman** [comment]
> The problem occurs when it's not _their_ code that needs fixing. It's the development libraries (e.g. boost) their applications build against, which are preferably provided by package manger versions.

Apologies if there were details elsewhere that I missed, but the issue is fixed in Boost. So if a user has a package manager which provides Clang 20 (releases mid 2025), wouldn't that package manager also already be providing the Boost (released in mid-2024) with the fix in it?

**Comment by RJVB** [comment]
On Wednesday December 04 2024 06:35:41 Aaron Ballman wrote:

>The Linux kernel itself, FreeBSD, Darwin, others...

Is there even C++ code in the Linux kernel? I did try to build one with clang, but quickly discovered that wasn't a good idea (IIRC it ended up not having support for dynamically loaded extensions).

**Comment by mborgerding** [comment]
RedHat Enterprise Linux 9.5 is the newest. It ships with boost 1.75. The bug was not fixed until 1.81
If someone is building up a system months from now they may grab the latest clang or intel compiler based thereon and expect it to work.  You might say the fault is RedHat's outdated versions. But right or wrong, compilers are expected to be backwards compatible.  If you violate that condition too often, you'll lose the user.

I'm ambivalent which way you go.  I appreciate that any pains from a hard error would be limited in scope and time; it would be cleaner in the long run to just "rip off the bandage".  But it is still pain.  Pain that could be avoided with either a workaround or limiting the hard error to >= c++17.
I appreciate the modularity and exacting nature of llvm/clang. It finds good diagnostics above what we get from g++. We'll continue to use it as a regression testing platform no matter what our customers do. _But_ intel compilers have historically been a PITA to support.  It would sometimes seem as if there was an evil genie (or lawyer) reading through the standard, carefully looking for new and inventive ways to break existing code.  Now that intel has switched over to LLVM, they inherit your decisions in this regard.  How do you want to proceed?

On a related note, does anyone know how to tell upon which version of LLVM a particular intel icpx is based?

**Comment by AaronBallman** [comment]
> Is there even C++ code in the Linux kernel?

I'm not certain where this discussion landed: https://www.phoronix.com/news/CPP-Linux-Kernel-2024-Discuss

> I did try to build one with clang, but quickly discovered that wasn't a good idea (IIRC it ended up not having support for dynamically loaded extensions).

You may want to check out https://www.kernel.org/doc/html/latest/kbuild/llvm.html

**Comment by AaronBallman** [comment]
> RedHat Enterprise Linux 9.5 is the newest. It ships with boost 1.75. The bug was not fixed until 1.81 If someone is building up a system months from now they may grab the latest clang or intel compiler based thereon and expect it to work. You might say the fault is RedHat's outdated versions. But right or wrong, compilers are expected to be backwards compatible. If you violate that condition too often, you'll lose the user.

Thanks, that's helpful!

> I'm ambivalent which way you go. I appreciate that any pains from a hard error would be limited in scope and time; it would be cleaner in the long run to just "rip off the bandage". But it is still pain. Pain that could be avoided with either a workaround or limiting the hard error to >= c++17. I appreciate the modularity and exacting nature of llvm/clang. It finds good diagnostics above what we get from g++. We'll continue to use it as a regression testing platform no matter what our customers do. _But_ intel compilers have historically been a PITA to support. It would sometimes seem as if there was an evil genie (or lawyer) reading through the standard, carefully looking for new and inventive ways to break existing code. Now that intel has switched over to LLVM, they inherit your decisions in this regard. How do you want to proceed?

Intel's downstream can (and does) carry patches to deviate from the upstream behavior, as with any of the other downstream consumers of Clang. In upstream, we try to pick behaviors that will minimize the need for downstreams to patch, so downstreams still do impact the decisions we make.

**Comment by RJVB** [comment]
On Wednesday December 04 2024 07:56:31 Aaron Ballman wrote:
>You may want to check out https://www.kernel.org/doc/html/latest/kbuild/llvm.html

Thanks, pretty certain I did, but in the end the question is moot on the hardware I use. As I observed elsewhere in this ticket, current/recent GCC compilers are faster and less resource hungry than even an old clang like 12.0.1 (the newest I currently have on the Linux in question). Given that neither compiler wins across the board in terms of performance of the produced code (not surprisingly) the choice makes itself for me, esp. for large projects (including building the compilers themselves, where GCC is also a clear winner).

**Comment by mborgerding** [comment]
> anyone know how to tell upon which version of LLVM a particular intel icpx is based?

I did not easily find anything official, but from fiddling with [godbolt](https://godbolt.org/z/4hdT6crev),
 I infer clang major.minor maps to icpx version as follows
19.0 --> 2024.2.0 up to at least 2025.0.0
18.0 --> 2024.1.0
17.0 --> 2023.2.1 through 2024.0.0
16.0 --> 2023.0.0 through 2023.1.0
...

**Comment by AaronBallman** [comment]
> (There is an interesting question there of whether this diagnostic should be tied to `-fstrict-enums` or not. Currently it's not.)

Related, Jonathan filed this issue in GCC: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=117963

**Comment by MaxEW707** [comment]
I am a little confused how the enum of unfixed type DR in C++17 and how that relates to C interoperability especially on Win32 APIs.
I ran into this error a couple of weeks ago. I try to keep up with the C and C++ std but time isn't always available so forgive me if anything I say below is incorrect.

On MSVC all enums of unfixed types have int as the underlying type. I am not worrying about the MSVC `/Zc:enumTypes` option here as it isn't used in shipping code since it breaks a lot of Win32.

A lot of Win32/DirectX/etc APIs use enums for constants instead of macros like POSIX.
The Windows SDK headers tend to add enum values over time but code in large projects need to build/ship against older Windows SDK headers. It is common to use runtime detection and your own constants to work around needing to ship against older Windows SDK headers especially in shared code that is used across different shipping products with different minimum specs.

Take the [HeapQueryInformation](https://learn.microsoft.com/en-us/windows/win32/api/heapapi/nf-heapapi-heapqueryinformation) Win32 API which uses the `HEAP_INFORMATION_CLASS` enum.

On Windows SDK 22621 that enum is defined as
```
typedef enum _HEAP_INFORMATION_CLASS {
    HeapCompatibilityInformation = 0,
    HeapEnableTerminationOnCorruption = 1

#if ((NTDDI_VERSION > NTDDI_WINBLUE) || \
     (NTDDI_VERSION == NTDDI_WINBLUE && defined(WINBLUE_KBSPRING14)))
    ,
    HeapOptimizeResources = 3
#endif
    ,
    HeapTag = 7
} HEAP_INFORMATION_CLASS;
```

On Windows SDK 19041 that enum is defined as
```
typedef enum _HEAP_INFORMATION_CLASS {
    HeapCompatibilityInformation = 0,
    HeapEnableTerminationOnCorruption = 1

#if ((NTDDI_VERSION > NTDDI_WINBLUE) || \
     (NTDDI_VERSION == NTDDI_WINBLUE && defined(WINBLUE_KBSPRING14)))
    ,
    HeapOptimizeResources = 3
#endif
} HEAP_INFORMATION_CLASS;
```

Having code such as
```
static constexpr HEAP_INFORMATION_CLASS kHeapTag = HEAP_INFORMATION_CLASS(7);

// Do runtime detection or just ignore a failure of `ERROR_INVALID_PARAMETER` since HeapTag is a debugging API
HeapQueryInformat(..., kHeapTag, ...);
```
now fails to compile which I understand the rationale for from the DR in C++17.

However from my understanding C allows any valid value that is representable by the underlying unfixed enum type.
Given the definition of `HEAP_INFORMATION_CLASS` above in `19041` and assuming MSVC compat so the enum is an int; in C any int value is valid but in C++17 only the values of 0-3 are valid.

```
// UB in C++17 as far as I understand when building with the definition above in 19041
HeapQueryInformat(...,  HEAP_INFORMATION_CLASS(7), ...);
```

If my understanding is correct one of the following would have to happen
- Win32 headers need to be updated to make enum decls have a fixed type in C++, 'enum HEAP_INFORMATION_CLASS : int`.
- MSVC may not follow this part of the spec and thus clang in MSVC compat mode shouldn't either. Don't treat this as UB for MSVC.

In my toy example above I could just check `WDK_NTDDI_VERSION` or other NTDDI macros from sdkddkver.h and not try to use `HeapTag` if building against older headers even if we may be running on a system that has such support but I just needed a small example.

I need to think a bit more about MSVC compat in clang here. Maybe we tie this to `-fstrict-enums` when in MSVC compat mode?
